Add a C++ Interoperability Vision Document (#60501)
apple/swift: docs/CppInteroperability/ForwardVision.md

“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story; however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and therefore a different vision document. [The vision for reverse interoperability is being developed here.](https://github.com/apple/swift/pull/61256)

This document is a prospective feature vision document, as described in the [draft review management guidelines](https://github.com/rjmccall/swift-evolution/blob/057b2383102f34c3d0f5b257f82bba0f5b94683d/review_management.md#future-directions-and-roadmaps) of the Swift evolution process.  It has not yet been approved by the Language Workgroup.

> ## Goals

Reviewer: Perhaps I've missed it, but do we have (rough) goals around what % of C++ APIs we want to be able to support basically "out of the box"/with minimal annotation burden?

> Safety is a top priority for the Swift programming language, which creates a tension with C++. While Swift enforces strong rules around things like memory safety, mutability, and nullability, C++ largely makes the programmer responsible for handling them correctly, on the pain of undefined behavior. Simply using C++ APIs should not completely undermine Swift's lanaguage guarantees, especially guarantees around safety. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the necessary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer. As a last resort, Swift should make an API unavailable if there's no reasonable path to a sufficiently safe Swift interface for it.

Author: Unfortunately, I think the nature of C++ is that there will be unsafe corner cases. There is always a trade-off between usability and safety, and we are already leaning heavily towards the safety side of things. This originally read "must," but we decided to weaken it, because otherwise we'd simply not import too many APIs (or break compatibility with Objective-C, for which this guarantee is not there). I think it's OK for this to be a bit subjective.

> Safety is a top priority for the Swift programming language, which creates a tension with C++. While Swift enforces strong rules around things like memory safety, mutability, and nullability, C++ largely makes the programmer responsible for handling them correctly, on the pain of undefined behavior. Simply using C++ APIs should not completely undermine Swift's lanaguage guarantees, especially guarantees around safety. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the necessary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer. As a last resort, Swift should make an API unavailable if there's no reasonable path to a sufficiently safe Swift interface for it.

Reviewer: > I think the nature of C++ is that there will be unsafe corner cases

I don't think any differently. 🙂 But IMO it should not be possible in _safe_ Swift to introduce unsafety that isn't due to the nature of the underlying C++ API itself. Would it be too restrictive to commit to something like "APIs which cannot be imported without introducing _additional_ safety concerns on the Swift side should be imported with an explicit indicator of that unsafety (such as `Unsafe` in the name)"?

> Safety is a top priority for the Swift programming language, which creates a tension with C++. While Swift enforces strong rules around things like memory safety, mutability, and nullability, C++ largely makes the programmer responsible for handling them correctly, on the pain of undefined behavior. Simply using C++ APIs should not completely undermine Swift's lanaguage guarantees, especially guarantees around safety. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the necessary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer. As a last resort, Swift should make an API unavailable if there's no reasonable path to a sufficiently safe Swift interface for it.

Author: I think the two big issues with adding unsafe to potentially unsafe APIs are 1) 90% of the time the APIs are safe, so "unsafe" starts to lose it's meaning. I think we should try to reserve this for the cases where we are more confident that it is actually an unsafe API. 2) this potentially breaks compatibility with Objective-C, or at least makes it more difficult to migrate bridging headers. Looking at the string example from one of the Swift compiler bridging headers:

```
BridgedIdentifier
SwiftASTContext_getIdentifier(
    void *ctx, const uint8_t * _Nullable str, long len);
```

This would definitely be imported as an "Unsafe" API (without more information). Maybe we want to enable C++ interop though, so we can start replacing these "Bridged" types with C++ types. But let's say we want to start using C++ interop incrementally, we can't update the whole bridging header in one go. This is where the issue comes in: if we start re-naming functions like this, it will prevent people from adopting C++ interop incrementally (they'll have to go find-replace all of these function names as soon as they enable C++ interop). Does that make sense?

> Safety is a top priority for the Swift programming language, which creates a tension with C++. While Swift enforces strong rules around things like memory safety, mutability, and nullability, C++ largely makes the programmer responsible for handling them correctly, on the pain of undefined behavior. Simply using C++ APIs should not completely undermine Swift's lanaguage guarantees, especially guarantees around safety. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the necessary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer. As a last resort, Swift should make an API unavailable if there's no reasonable path to a sufficiently safe Swift interface for it.

Reviewer: Hm, I don't think I'm suggesting that this function would necessarily have to change to be called `unsafe_SwiftASTContext_getIdentifier` or something—this API would _already_ be imported to Swift with 'unsafe' sigils on its `ctx` and `str` parameters.

Further, I'm not saying that _every_ potentially unsafe C++ API should be imported to Swift with 'unsafe' markings, only those for which importing to Swift introduces _new_ forms of unsafety that would not be possible to achieve in C++. I.e., the APIs that we cannot guarantee are "[at least as] safe to use from Swift [as] they would be in C++." Is this `SwiftASTContext_getIdentifier` an API that would fall into that bucket, you think?

## Lifetime and safety of owned types

Value types that own memory do not natively exist in Swift today. Because Swift was not built around this kind of value type, Swift handles the lifetime of value types in a subtly different way from C++. This subtle difference makes a naive mapping of C++ value types that own memory to native Swift value types extremely dangerous, especially when dealing with projections of owned storage. Let’s look at an example Swift program that naively imports some owned type and returns a projections of it:

```
var v = vector(1)
let start = v.begin()
doSomething(start)
fixLifetime(v)
```

To understand the problem with this code, the following snippet highlights where an implicit copy is created and destroyed:

```
var v = vector(1)
let copy = copy(v)
let start = copy.begin()
destroy(copy)
doSomething(start)
fixLifetime(v)
```

Here, because Swift copies `v` into a temporary with a tight lifetime before the call to `begin`, `v` projects a dangling reference. This is an example of how subtly different lifetime models make using C++ types from Swift hard, if their semantics aren’t understood by the compiler.

> To make these APIs safe and usable, Swift cannot import unsafe projections of types that own memory, because they don’t fit the Swift model. Instead, the Swift compiler can try to infer what, semantically, the API is trying to do, or the library author can provide this information via annotations. In this case, the Swift compiler can infer that begin returns an iterator, which Swift can represent through the existing, safe Swift iterator interface. In the example above, “start” is a pointer type. Using this pointer returned by the “begin” method is unsafe, but the type of start itself is not unsafe. In other words, safety restrictions need not be applied to pointer types themselves but rather their unsafe uses.

Reviewer: Is it counter to our goals to do things like _change_ lifetime behaviors for imported C++ types to make them behave a little more C++ like? I.e., use lexical scoping rules for imported C++ types here.

Author: Let me start off by saying I think that's a good idea (in this specific case). However, more generally, I think that's a slippery slope. There's always a tradeoff when making imported APIs behave less like "Swift". It may make it easier to use the APIs, but as we lose Swift semantics, which creates cognitive burden for the programmer. Drifting away from Swift semantics also means programers might not get a lot of the benefit of using Swift. We don't want this to feel like C++ with different syntax, we want it to feel like Swift. 

That is all to say, I think we need to be very selective in where we do this.

Reviewer: Yeah, I see the difficulty. On the one hand, working with C++ APIs is going to require _some_ knowledge of C++ semantics. Imported Objective-C classes today exhibit behavior that cannot really be explained in terms of Swift semantics and I expect it will be the same for imported C++ APIs. On the other hand there are also invariants of Swift that we will likely want to make sure get preserved even for imported C++ APIs.

This model cannot be adopted in Swift, however, because the the same lexical lifetime model does not exist. Further, projections of owned types are completely foreign concept in Swift, meaning users aren’t familiar with programming in terms of this lexical model, and may not be aware of the added (implicit) constraints (that is, when objects are destroyed). Swift’s language model is such that returning projections from a copied value, even in smaller lexical scope, should be safe. In order to allow projections of owned types, this assumption must be broken, or C++ interoperability must take advantage of Swift ownership features to associate the lifetime of the projection to the source.

The following example highlights the case described above:

```
func getCString(str: std.string) -> UnsafePointer<CChar> { str.c_str() }
```

> The above function returns a dangling reference to `str`‘s inner storage. In C++, it is assumed that the programmer understands this is a bug, and generally would be expected to take `str` by reference. This is not the case in Swift. To represent this idiomatically in Swift, the lifetimes must be associated through a projection. Using the tools provided in the ownership manifesto this would mean yielding the value returned by `c_str` out of a [generalized accessor](https://github.com/apple/swift/blob/main/docs/OwnershipManifesto.md#generalized-accessors)(resulting in an error when the pointer is returned).

Reviewer: What is "this" in "This is not the case in Swift"? This code would raise eyebrows for me in Swift even if it weren't using a C++ type.

Because of the difference in idioms between the two languages, and because of the safety concerns when exposing certain APIs to Swift, a C++ API might look quite different in Swift than it does in C++. It is a goal of C++ interoperability to provide a clear, well-defined mapping for whether and how APIs are imported into Swift. Users should be able to read the C++ interoperability documentation to have a good idea of how much of their API will be able to imported and what it will look like. Swift should also provide tools for inspecting what a C++ API will look like in Swift, and these tools should call out notable parts of the API that were not imported.

Finally, Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 

## The approach

> 

Reviewer 2: I think you've got the right ideas for what you want to say here, but it's missing a cohesive narrative.  Let me suggest how you can rework the first few paragraphs of this section.

```suggestion
The example above shows how the basic tools provided by C++ are often used in idiomatically different ways.  This poses a conundrum for importing C++ APIs automatically into Swift, and it's bigger than just reference vs. value semantics for classes.  A more fundamental place to see this is with memory management.

In Objective-C, it's fairly straightforward for ARC to ensure that data is valid when it's used.  Nearly all data in Objective-C is represented with either a fundamental type like `double` or `BOOL` or a managed object reference type like `NSString*`.  Fundamental values can be freely copied around, and managed object references can be conservatively retained to extend their lifetime.  It's rare to work with unmanaged C pointers in Objective-C, and even rarer to work with unmanaged pointer that have a dependency on a managed object (such as [`NSData`'s `-bytes` method](https://developer.apple.com/documentation/foundation/nsdata/1410616-bytes?language=objc)).  These latter cases can be treated as special cases and dealt with individually.

In contrast, it's very common for C++ APIs to work with unmanaged pointers, references, and views into other objects.  The lifetime rules for using these correctly are often varied, maybe even unique to an API.  As an example, consider three values: a value of type `std::vector<std::string>`, a reference returned from that vector's subscript operator, and an iterator returned from that vector's `begin()` method.  On a primitive level, these values look similar: they are all either pointers or class objects containing pointers.  But each has its own semantics and expected use, and these differences are not completely explicit in the source.  The vector is a value type that can be copied, but copies can be expensive, and iterators and references into the vector are valid only for a specific copy.  The subscript result is a mutable projection of a specific element, dependent on the vector for validity; but that element value can be copied out of the reference to get an independent value, and that is often how the subscript operator is used.  The iterator is also a projection, dependent on the vector for validity, but it must be used in conjunction with other iterators or with the vector itself in certain careful ways, and some operations will invalidate it completely.

So there is a conundrum where superficially similar language structures in C++ are used to express idiomatic patterns that are vastly different in their impact.  The only viable approach for addressing this conundrum is to pick off these patterns one at a time.  Swift's C++ importer will apply standard rules to sort C++ APIs into different idioms that it knows how to recognize.  The programmer will also be able to intervene by adding annotations to guide the importer.  If the standard rules don't allow the API to be imported, or if the API has an annotation that it doesn't actually satisfy the rules for, the importer will make the API unavailable, leaving enough breadcrumbs to tell the programmer what went wrong if they try to use it.

Consider how this applies to the `std::vector` example.  `std::vector<std::string>` maps over well as a Swift value type.  Its subscript operator can be imported as a Swift `subscript`, and the importer can take advantage of the fact that it returns a reference to allow elements to be efficiently borrowed.  And while C++ iterators in general pose serious lifetime safety problems in Swift, Swift can recognize the common `begin()`/`end()` idiom and import it as a safe Swift iterator that encapsulates the unsafety internally.
```

The detailed approach for importing C++ APIs into Swift follows from the goals above. Much like the example above picked off vairous API patterns from the mess of APIs that `std::vector` provides, C++ interoperability will need to pick off specific API patterns from the mess of all C++ APIs. For example, `std::vector` itself maps well to a Swift value type. The subscript operator can be bridged to a Swift subscript with a getter and a setter. And with some bridging, the `begin` method can be mapped to a safe Swift iterator protocol. The Swift compiler will know about many possible C++ API patterns. If a C++ API has semantic annotations telling Swift that it follows a certain pattern, Swift will try to create a Swift interface for it following the rules of that pattern. In the absence of those annotations, Swift will try to use huristics to recognize an appropriate pattern. If this fails, Swift will make the API unavailable.

> One of the most common uses of this "API patterns" concept concerns the import of types. Swift types fall into two categories: value types and reference types. Copying a value of a reference type produces a new reference to the same underlying object, similar to an intrusive `std::shared_ptr` in C++ or a class type in Java. Copying a value of a value type recursively copies the components of the type, similar to the behavior of a struct in C or the default behavior of a class type in C++. Furthermore, both kinds of types must always be copyable, although there are plans in the works to allow types to restrict this.

Reviewer 2: Okay, so we pick up here with this paragraph after the suggestion above.  I'll provide more feedback starting from here tomorrow.

The "API patterns" model picks off various C++ APIs that Swift can handle in a specific, idomatic way such as self contained value types, reference types, singletons, iterators, generalized accessors, and so on.

### Reference types

Like Swift types, Objective-C types fall into two categories, which makes importing them easy: structs are imported as value types and Objective-C classes are imported as Swift class types. The same is not true for C++. As discussed above, there is no clear idiom for defining reference types in C++. All types in C++ are declared in the same way, but some types have reference semantics and others don’t. To be able to express reference types in Swift natively, users must annotate their reference types as such, which will tell the compiler to import them as Swift classes (which have the same semantics).

> When importing a reference type, there are three main sub-categories that reference type can fall into: 1) **immortal** reference types are “intentionally leaked” and live for the duration of the program 2) **manually managed** reference types are constructed and destroyed manually by the programmer, these are unsafe reference types and should maybe be renamed accordingly 3) **automatically managed** reference types provide retain and release operations which are used to safely destroy a value of the type after its last use. [(Examples below.)](## Examples and Definitions)

Reviewer 2: ```suggestion
There are several common patterns for reference types in C++. Swift's C++ interoperability should support at least these three:
- **Immortal** reference types are not designed to be managed individually by the program. Objects of these types are allocated and then intentionally "leaked" without tracking their uses.

  Sometimes these objects are not truly immortal: for example, they may be arena-allocated, with an expectation that they will only be referenced from other objects within the arena. Nonetheless, they aren't expected to be individually managed.
- **Manually managed** reference types are constructed and destroyed manually by the programmer.  Usually they're uniquely referenced and there's some sort of explicit operation to destroy them, either the standard `delete` operator or a specific method.
- **Automatically managed** reference types are reference-counted with retain and release operations.
[Examples of each of these are given below.](## Examples and Definitions)
```

I have to say that I think these last two are poorly-chosen names.  I think you're talking about unique reference types and shared reference types, and there's an orthogonal dimension of whether or not the API expects the programmer to manage ownership explicitly.

  std::array<std::string, 32> names;
  std::array<std::string, 32> places;
  // ...
  
  StatefulObject(const StatefulObject&) = delete;
  StatefulObject() = delete;

  StatefulObject *create() { return new StatefulObject(); }
};
```

Because this object is so expensive to copy, the programmer decided to delete the copy constructor. The programmer also decided that this object should be allocated on the heap, so they decided to delete the default constructor, and provide a create method in its place. 

In Swift, this `StatefulObject` should be imported as a reference type, as it has reference semantics.

**API Incorrectly Using Reference Types**

Here someone has written an API that uses `StatefulObject` as a value type.

```
StatefulObject makeAppState();
```

This will invoke a copy of `StatefulObject` which violates the semantics that the API was written with. To be useable from Swift, this API needs to be updated to pass the object indirectly (by reference):

```
StatefulObject *makeAppState(); // OK
const StatefulObject *makeAppState(); // OK
StatefulObject &makeAppState(); // OK
const StatefulObject &makeAppState(); // OK
```

**Immortal Reference Types**

> Instances of StatefulObject above are manually managed by the programmer, they create it with the create method and are responsible for destroying it once it is no longer needed. However, some reference types need to exist for the duration of the program, these references types are known as “immortal.” Examples of these immortal reference types might be pool allocators or app contexts. Let’s look at a GameContext object which allocates (and owns) various game elements:

Reviewer 3: formatting nit
```suggestion
Instances of `StatefulObject` above are manually managed by the programmer, they create it with the create method and are responsible for destroying it once it is no longer needed. However, some reference types need to exist for the duration of the program, these references types are known as “immortal.” Examples of these immortal reference types might be pool allocators or app contexts. Let’s look at a GameContext object which allocates (and owns) various game elements:
```

```
struct GameContext {
  // ...
  
  GameContext(const GameContext&) = delete;

  Player *createPlayer();
  Scene  *createScene();
  Camera *createCamera();
};
```

> Here the GameContext is meant to last for the entire game as a global allocator/state. Because the context will never be deallocated, it is known as an “immortal reference type” and the Swift compiler can make certain assumptions about it. 

Reviewer 3: formatting nit
```suggestion
Here the `GameContext` is meant to last for the entire game as a global allocator/state. Because the context will never be deallocated, it is known as an “immortal reference type” and the Swift compiler can make certain assumptions about it. 
```
